File: leetcode/17_Letter_Combinations_of_a_Phone_Number.py
import unittest
from typing import List

class Solution:
    def _rec(self, digits, n, pos, letters, mapping, combinations):
        for ch in mapping[digits[pos]]:
            letters.append(ch)
            if pos < n - 1:
                self._rec(digits, n, pos + 1, letters, mapping, combinations)
            else:
                combinations.append("".join(letters))
            letters.pop()

    def letterCombinations(self, digits: str) -> List[str]:
        n = len(digits)
        mapping = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl", "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}
        combinations = []
        if n > 0:
            self._rec(digits, n, 0, [], mapping, combinations)
        return combinations

class TestLetterCombinations(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().letterCombinations(""), [])    

    def test_2(self):
        self.assertEqual(Solution().letterCombinations("2"), ["a", "b", "c"])            

    def test_3(self):
        self.assertEqual(Solution().letterCombinations("23"), ["ad","ae","af","bd","be","bf","cd","ce","cf"])                    

File: leetcode/17_Letter_Combinations_of_a_Phone_Number.rs
struct Solution {
}

use std::collections::HashMap;

macro_rules! hashmap {
    ($( $key: expr => $val: expr ),*) => {{
         let mut map = HashMap::new();
         $( map.insert($key, $val); )*
         map
    }}
}


impl Solution {

    fn _rec(digits: &Vec<char>, n: usize, pos: usize, combination: &mut String, mapping: &HashMap<char, Vec<char>>, combinations: &mut Vec<String>) {
        for ch in &mapping[&digits[pos]] {
            combination.push(*ch);
            if pos + 1 < n {
                Self::_rec(digits, n, pos + 1, combination, mapping, combinations);
            } else {
                combinations.push(combination.clone());
            }
            combination.pop();
        }
    }

    pub fn letter_combinations(digits: String) -> Vec<String> {
        let digits: Vec<char> = digits.chars().collect();
        let n = digits.len();
        let mapping = hashmap![
            '2' => vec!['a', 'b', 'c'], 
            '3' => vec!['d', 'e', 'f'], 
            '4' => vec!['g', 'h', 'i'], 
            '5' => vec!['j', 'k', 'l'], 
            '6' => vec!['m', 'n', 'o'], 
            '7' => vec!['p', 'q', 'r', 's'], 
            '8' => vec!['t', 'u', 'v'], 
            '9' => vec!['w', 'x', 'y', 'z']];
        let mut combinations = Vec::<String>::new();
        let mut combination = String::new();
        if n > 0 {
            Self::_rec(&digits, n, 0, &mut combination, &mapping, &mut combinations);
        }
        combinations
    }
}

#[cfg(test)]
mod letter_combinations_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::letter_combinations(String::from("")), Vec::<String>::new());
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::letter_combinations(String::from("2")), vec![String::from("a"), String::from("b"), String::from("c")]);
    }
}

